Discovery documentation update [7926] (#100)
eProsima/Fast-RTPS-docs: docs/use-cases.rst

|    :end-before: <!--><-->                              |    :end-before: <!--><-->                              |
+--------------------------------------------------------+--------------------------------------------------------+

.. _discovery_server_redundancy_scenario_setup:

UDPv4 redundancy example
^^^^^^^^^^^^^^^^^^^^^^^^

The :ref:`above example <discovery_server_major_scenario_setup>` presents a *single point of failure*, that is, if the

> *server* falls there is no discovery. In order to prevent this, several servers could be linked to a *client*. By doing

Author: ```suggestion
*server* fails there is no discovery. In order to prevent this, several servers could be linked to a *client*. By doing
```

Reviewer: Fall and fail imply different things in this context but, given that I speak of failure later on, let's buy it.

> this, a discovery failure only takes place if *all servers* fail, a more unlikely event.

Author: ```suggestion
this, a discovery failure only takes place if *all servers* fail, which is a more unlikely event.
```

Reviewer: ok

We must make sure that the each server has a unique **Prefix** and *unicast address*. In the example below we have

> chosen:

Author: Reported speech

Reviewer: lets say:
> The following values have been chosen in order to assure each server has a unique **Prefix** and *unicast address*:

.. csv-table::
    :header: "Prefix", "UDPv4 address"
    :widths: 20,100

    75.63.2D.73.76.72.63.6C.6E.74.2D.31, "192.168.10.57:56542"
    75.63.2D.73.76.72.63.6C.6E.74.2D.32, "192.168.10.60:56543"

.. image:: ds_redundancy.png
    :align: center
    :width: 75%

Note that several *servers* can share the same *IP address* but the port number should be different. Likewise several

> *servers* can share the same port if its *IP address* is different.

Author: ```suggestion
Note that several *servers* can share the same *IP address* but their port numbers should be different. Likewise, several
*servers* can share the same port if their *IP addresses* are different.
```

Reviewer: ok

|    :end-before: //!--                                  |    :end-before: //!--                                  |
+--------------------------------------------------------+--------------------------------------------------------+
| **XML**                                                | **XML**                                                |
+--------------------------------------------------------+--------------------------------------------------------+
| .. literalinclude:: ../code/XMLTester.xml              | .. literalinclude:: ../code/XMLTester.xml              |
|    :language: xml                                      |    :language: xml                                      |
|    :start-after: <!-->CONF_DS_RDNCY_SCENARIO_SERVER<-->|    :start-after: <!-->CONF_DS_RDNCY_SCENARIO_CLIENT<-->|
|    :end-before: <!--><-->                              |    :end-before: <!--><-->                              |
+--------------------------------------------------------+--------------------------------------------------------+

.. _discovery_server_persistency_scenario_setup:

> UDPv4 persistency example

Author: In this example:

* Use reported speech, as in the rest of the sections
* I think the text needs to be more "story telling". The sentences do not feel connected to each other (in fact each one is a paragraph).
* A bit of context, you gotta think that the person reading this may not read anything else.
* I miss snippets, I thing we should have them in all the examples. Again, since this is a use-case, the read may not have read the previous sections.

Reviewer: When I wrote it I was pondering if introduce cumbersome source and xml snippets for the partitioning because only the first case has an associated snippet. I thought those snippets will be the same thing again and again and confuse more than clarify. But maybe you're right, let's go for it!

- *clients* receive this information from its *servers*.
- *servers* receive this information from its *clients*.

Given that *servers* used to have many *clients* associated, this is a lengthy process. In case of *server* failure we
may be interested in speed up this process when the *server* restarts.

Keep the discovery information in a file synchronize with the *server*'s record fulfills the goal. In order to enable
this we must just specify the :ref:`discovery protocol <discovery_protocol>` as **BACKUP**.

Once the *server* is created it generates a *server-<GUIDPREFIX>.db* (*exempli gratia
server-73-65-72-76-65-72-63-6C-69-65-6E-74.db*) on its process working directory.

If we want to start anew we must remove the file in order to prevent old discovery info from being loaded.

.. _discovery_server_partitioning_setup:

UDPv4 partitioning using servers
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

*Server* association can be seen as another isolation mechanism besides :ref:`domains <listening_locators>` and
:ref:`partitions <partitions>`. *Clients* that do not share a *server* cannot see each other and belong to isolated

> networks. In order to connect server isolated networks we can:

Author: ```suggestion
server networks. In order to connect server isolated networks we can:
```

Reviewer: ok

    1. Connect each *client* to both *servers*.
    2. Connect one *server* to the other.

>     3. Create a new *server* linked to the *servers* the clients are connected to.

Author: ```suggestion
    3. Create a new *server* linked to the *servers* to which the clients are connected.
```

Reviewer: ok

Options 1 and 2 can only be implemented by modifying attributes or XML configuration files beforehand. In this regard
they match the domain and partition strategy. Option 3 can be implemented at runtime, that is, when the isolated

> networks are already up and running.

Author: I'd provide snippet examples here

Reviewer: Ok, The first case is already covered above but I can introduce the others.

Author: :+1: You could make a link to the first one then

connections to those strictly necessary.
As explained in the :ref:`discovery` section, all Fast-RTPS discovery mechanisms consist of two steps: PDP and EDP.

> 

Author: ```suggestion
```

Reviewer: ok
